### PR TITLE
[FEATURE] `DatasourceStore`

### PR DESCRIPTION
**Pull request by cdkini**
Changes proposed in this pull request:
- New `DatasourceStore` that mirrors existing Store logic
- Modification of `GeCloudStoreBackend` to allow for new resource type.
- Tests to ensure proper serialization.

JIRA: https://superconductive.atlassian.net/browse/GREAT-910

### Definition of Done
Please delete options that are not relevant.

- [x] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/docs/contributing/style_guides/code_style)
- [x] I have performed a [self-review](https://docs.greatexpectations.io/docs/contributing/contributing_checklist) of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added [unit tests](https://docs.greatexpectations.io/docs/contributing/contributing_test#writing-unit-and-integration-tests) where applicable and made sure that new and existing tests are passing.
- [x] I have run any local integration tests and made sure that nothing is broken.

Thank you for submitting!

### CODE REVIEW

File: tests/data_context/store/test_datasource_store.py [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+    empty_datasource_store: DatasourceStore,
+    datasource_config: DatasourceConfig,
+    datasource_key: ConfigurationIdentifier,
+):
+    assert len(empty_datasource_store.list_keys()) == 0
+    empty_datasource_store.set(key=datasource_key, value=datasource_config)
+    assert len(empty_datasource_store.list_keys()) == 1
+
+
+def test_datasource_store_integration(
```
Comment (anthonyburdi) [inline tests/data_context/store/test_datasource_store.py]: It might make sense to mark this as an integration test so we can use flags to exclude/include running it in CI pipelines as we organize our tests a bit more.
```suggestion
@pytest.mark.integration
def test_datasource_store_integration(
```

**Review by anthonyburdi — APPROVED** [review]
LGTM! One minor non blocking suggestion.

File: tests/data_context/store/test_datasource_store.py (hunk 1 above)
Comment (cdkini) [reply tests/data_context/store/test_datasource_store.py]: I love that! Great call, I've committed your suggestion.

File: great_expectations/data_context/store/datasource_store.py [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+
+class DatasourceStore(ConfigurationStore):
+    """
+    A DatasourceStore manages Datasources for the DataContext.
+    """
+
+    _configuration_class = DatasourceConfig
+
+    def serialization_self_check(self, pretty_print: bool) -> None:
```
Comment (donaldheppner) [inline great_expectations/data_context/store/datasource_store.py]: why is this here? This is test code

File: great_expectations/data_context/types/base.py [hunk 3]
```diff
@@ -976,6 +891,108 @@ def make_datasource_config(self, data, **kwargs):
         return DatasourceConfig(**data)
 
 
+class DatasourceConfig(BaseYamlConfig):
```
Comment (donaldheppner) [inline great_expectations/data_context/types/base.py]: Given this type does not inherit from `BaseYamlConfig` today, can we avoid introducing it? Eventually we don't want any of our config objects to inherit from BaseYamlConfig

**Review by donaldheppner — CHANGES_REQUESTED** [review]

File: great_expectations/data_context/store/datasource_store.py (hunk 2 above)
Comment (cdkini) [reply great_expectations/data_context/store/datasource_store.py]: that's a very good question 😄 I just followed the existing pattern but I do agree that this shouldn't be in our source code. I was under the impression that this was a necessary abstract method to implement to fulfill the parent's contract but I think we can forgo this.

I'll move this logic over to a test.

**Comment by sdebruyn** [comment]
Shouldn't this PR include some documentation about this new feature?

File: great_expectations/data_context/data_context/abstract_data_context.py [hunk 4]
```diff
@@ -11,4 +11,6 @@ class AbstractDataContext(ABC):
     TODO: eventually the dependency on ConfigPeer will be removed and this will become a pure ABC.
     """
 
-    pass
+    @abstractmethod
+    def _init_datasource_store(self) -> None:
+        raise NotImplementedError
```
Comment (cdkini) [inline great_expectations/data_context/data_context/abstract_data_context.py]: To be fleshed out in a subsequent PR

File: scripts/check_type_hint_coverage.py [hunk 5]
```diff
@@ -3,7 +3,7 @@
 from typing import Dict, List, Optional
 
 TYPE_HINT_ERROR_THRESHOLD: int = (
-    2548  # This number is to be reduced as we annotate more functions!
+    2547  # This number is to be reduced as we annotate more functions!
```
Comment (donaldheppner) [inline scripts/check_type_hint_coverage.py]: 🙌

File: tests/data_context/store/test_datasource_store.py [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+    empty_datasource_store: DatasourceStore, datasource_config: DatasourceConfig
+) -> None:
+    store: DatasourceStore = empty_datasource_store
+
+    key: StringKey = StringKey(key="my_datasource")
+    store.set(key=key, value=datasource_config)
+    res: DatasourceConfig = store.get(key=key)
+
+    assert isinstance(res, DatasourceConfig)
+    assert DatasourceConfigSchema().dump(res) == datasource_config.to_dict()
```
Comment (donaldheppner) [inline tests/data_context/store/test_datasource_store.py]: asserting the key here should be sufficient?

File: tests/data_context/store/test_datasource_store.py [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+    key: GeCloudIdentifier = GeCloudIdentifier(
+        resource_type="datasource", ge_cloud_id="foobarbaz"
+    )
+
+    with patch("requests.patch", autospec=True) as mock_patch:
+        type(mock_patch.return_value).status_code = PropertyMock(return_value=200)
+
+        store.set(key=key, value=datasource_config)
+
+        mock_patch.assert_called_with(
```
Comment (donaldheppner) [inline tests/data_context/store/test_datasource_store.py:120]: generally love this, can we leverage the datasource_config() fixture to avoid duplicating the payload here?

File: tests/data_context/store/test_datasource_store.py (hunk 6 above)
Comment (cdkini) [reply tests/data_context/store/test_datasource_store.py]: so unfortunately that doesn't seem to be the case - while both items are indeed `DatasourceConfig`, the fixture has dictionary attributes and the `res` obj has `Config` types.

Basically, `tripdata_monthly_configured` is of type `dict` in our fixture while it is of type `DataConnectoConfig` once parsed by the store.

Here's a Loom to elaborate: https://www.loom.com/share/983f45c656ce4157a8dd905ee085978d

File: tests/data_context/store/test_datasource_store.py (hunk 7 above)
Comment (cdkini) [reply tests/data_context/store/test_datasource_store.py:120]: that's a great call! done!

File: tests/data_context/store/test_datasource_store.py (hunk 6 above)
Comment (cdkini) [reply tests/data_context/store/test_datasource_store.py]: I also don't think that `DatasourceConfig` has implemented `__eq__` so it will compare by memory address instead of actual values.

File: tests/data_context/store/test_datasource_store.py [hunk 8]
```diff
@@ -124,30 +124,8 @@ def test_datasource_store_retrieval_cloud_mode(
...
-                                }
-                            },
-                            "execution_engine": {
-                                "module_name": "great_expectations.execution_engine",
-                                "class_name": "PandasExecutionEngine",
-                            },
-                            "class_name": "Datasource",
-                        },
-                        "organization_id": "bd20fead-2c31-4392-bcd1-f1e87ad5a79c",
+                        "datasource_config": datasource_config.to_dict(),
```
Comment (donaldheppner) [inline tests/data_context/store/test_datasource_store.py:127]: :heart::heart::heart::heart::heart::heart::heart:

**Comment by cdkini** [comment]
@sdebruyn apologies for the delayed response! This feature is very much under development and will still take some work before we're ready to publish it and present it to the larger community.

We'll be sure to keep you updated through Slack and our blog posts as things progress 😄

File: tests/data_context/store/test_datasource_store.py (hunk 6 above)
Comment (donaldheppner) [reply tests/data_context/store/test_datasource_store.py]: sorry for not being clear, I meant: could we assert key values for equality vs. the whole payload and still be testing what we intend to be

**Review by donaldheppner — APPROVED** [review]
